### ISSUE TITLE
Slow page edit in AdminThemUikit

### ISSUE THREAD

**Issue by matjazpotocnik** [processwire/processwire-issues#520]
In AdminThemeUikit the page edit is slow. It's related to the number of images/files on the page - higher the number of images, longer it takes the page to render. There must be some heavy duty lifting present in /wire/modules/AdminTheme/AdminThemeUikit/scripts/main.js as execution stops there for a few seconds, CPU goes up high and then other scripts are loaded and executed. See attached video, watch to the end.

![pageload](https://user-images.githubusercontent.com/7663992/37194963-228c25c6-2371-11e8-9495-0d5a2f8690e7.gif)

It looks like the problem is rather `initTags($(".InputfieldFileHasTags"));` in InputfieldFile.js, actually the call `$selects.selectize()` takes that long. If I use AdminThemeDefault or AdminThemeReno then page load is normal. If I disable tags on file/image fields then it's also ok. Don't know why it happens only on AdminThemUikit...

**Comment by ryancramerdesign** [comment]
It looks like once I get 30+ files uploaded, Selectize becomes very slow, whether it's init, or even just trying to add a tag. 

I'm not observing a slowdown in main.js, though I am in InputfieldFile.js in the selectize() call. Are you still seeing some issue in main.js?

**Comment by matjazpotocnik** [comment]
I have around 60 images on a page. I don't know why it gets stalled in main.js since selectize() is called in  InputfieldFile.js What bothers me more is why it works in AdmineThemeDefault (and probably in Reno)? Could it be some html markup ina uikit theme that makes selectize to work harder?

**Comment by ryancramerdesign** [comment]
Other than addition of uk-* class names, I don't think the markup is
different in any consequential way. There aren't any uk-classes added to
any part of the tags stuff. I'm thinking this likely has something to do
with JS events. I can see that Selectize is attaching a lot of
keyup/keydown events at the document level, so am taking a closer look at
those.

**Comment by ryancramerdesign** [comment]
@matjazpotocnik looks like this autoGrow() function call in Selectize.js is where the bottleneck is: https://github.com/selectize/selectize.js/blob/master/dist/js/standalone/selectize.js#L1311

**Comment by ryancramerdesign** [comment]
That autoGrow() function calls a measureString() function that gets called for every init/update, keydown, keyup and blur event in Selectize inputs:
https://github.com/selectize/selectize.js/blob/master/dist/js/standalone/selectize.js#L1004

That measureString() function is slow because it adds a `<test>` element to the DOM, populates text and styles to it, measures the width of it, and then deletes it... and it gets called a LOT. That just increases with every tags input, so it wasn't scaling well. Most likely Uikit 3 is monitoring the DOM for changes, so that is also getting triggered every time Selectize adds and deletes that `<test>` element to the document. This is probably why it was noticed in AdminThemeUikit and not in Default/Reno themes. Though I think it was also slowing down those other themes to some extent too.  

I fixed it by making that measureString() function keep its test element in the DOM rather than re-creating and deleting it every time. That solved the bottleneck, but I also added a cache to it, which I'm not sure is necessary, but figure it can't hurt.
